src/compiler/scala/tools/nsc/transform/Erasure.scala

```diff
@@ -805,6 +805,11 @@ abstract class Erasure extends InfoTransform
                 return result setType ErasedValueType(tref.sym, result.tpe)
 
             }
+          case tree @ Block(_, _: LabelDef) =>
+            // Avoid pushing the adaptation to a matchEnd label.
```

Comment:
it's not clear to me what "the adaptation" is that needs to be prevented

Reply:
I'll first try to explain here, and if that's OK I'll then expand in the comment:

The situation is as follows. `tree` is of the form
```scala
{
  stats
  matchEnd1(x: Foo) {
    x
  }
}
```
and the expected type `pt` is `Bar` (which is not a subtype of `Foo`).

Previously `super.typed1(adaptMember(tree), mode, pt)` would push `pt` to the `LabelDef`, which would result in
```scala
{
  stats
  (matchEnd1(x: Foo) {
    x
  }).asInstanceOf[Bar]
}
```
This is problematic for the back-end, as it's not a `Block(stats, LabelDef(...))` anymore, but a `Block(stats, TypeApply(LabelDef(...), ...))`. So it bails to the non-optimal codegen strategy.

By explicitly given `WildcardType` when recursing, what we receive for `super.typed1` is unchanged. And therefore, the later `adapt` inserts the `asInstanceOf` around the `Block`, as follows:
```scala
({
  stats
  matchEnd1(x: Foo) {
    x
  }
}).asInstanceOf[Bar]
```
and this is good, because it preserves the shape `Block(stats, labelDef(...))`.

Does that make sense?

Reply:
Sounds good, maybe something like "push adaptations out to the block to preserve patmat tree shapes" makes it more clear.